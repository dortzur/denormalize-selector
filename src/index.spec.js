import renorm from './';
import { normalize } from 'normalizr';
import stockFixture from '../fixtures/stock-fixure';
import { Schemas } from '../fixtures/schema-fixture';
import produce from 'immer';
import { decamelizeKeys } from 'humps';
const getStocks = (state) => state.stocks;

let state = null;

describe('renorm', () => {
  beforeAll(() => {
    const { result: stocks, entities } = normalize(
      stockFixture,
      Schemas.COMPANY_ARRAY
    );
    state = { stocks, entities };
  });

  it('denormalizes entities', () => {
    const getStocksSelector = renorm(getStocks, Schemas.COMPANY_ARRAY);
    const initialStocks = getStocksSelector(state);
    expect(initialStocks[0]).toMatchSnapshot();
    expect(initialStocks[2]).toMatchSnapshot();
    expect(initialStocks[15]).toMatchSnapshot();
    expect(initialStocks[30]).toMatchSnapshot();
  });

  it('replaces only changed denormalized entities', () => {
    const getStocksSelector = renorm(getStocks, Schemas.COMPANY_ARRAY);
    const initialStocks = getStocksSelector(state);
    const newState = produce(state, (draftState) => {
      draftState.entities.earnings['AAPL_QUARTER_1'].earnings = 15;
    });
    const newStocks = getStocksSelector(newState);
    const changedStocks = newStocks.filter(
      (stock, index) => stock !== initialStocks[index]
    );
    expect(changedStocks).toHaveLength(1);
    expect(changedStocks[0].id).toEqual('COMP_AAPL');
  });

  it('caches values correctly', () => {
    const getStocksSelector = renorm(getStocks, Schemas.COMPANY_ARRAY);
    getStocksSelector(state);
    expect(getStocksSelector.recomputations()).toEqual(1);
    getStocksSelector(state);
    expect(getStocksSelector.recomputations()).toEqual(1);
    state = { ...state, entities: { ...state.entities } };
    getStocksSelector(state);
    expect(getStocksSelector.recomputations()).toEqual(1);
    const stocks = getStocksSelector(state);
    state = produce(state, (draftState) => {
      draftState.stocks.push('COMP_AAPL');
      return draftState;
    });
    const changedStocks = getStocksSelector(state);

    expect(getStocksSelector.recomputations()).toEqual(2);
    expect(changedStocks).toHaveLength(stocks.length + 1);
    expect(
      changedStocks.filter((stock, index) => stock !== stocks[index])
    ).toHaveLength(1);
  });

  it('denormalizes a single entity', () => {
    const getAppleStock = renorm(() => 'AAPL', Schemas.STOCK);
    const appleStock = getAppleStock(state);
    getAppleStock(state);
    expect(appleStock).toMatchSnapshot();
    getAppleStock({ ...state });
    expect(getAppleStock.recomputations()).toEqual(1);

    const newState = produce(state, (draftState) => {
      draftState.entities.stocks['AAPL'].name = 'Apple';
    });
    getAppleStock(newState);
    expect(getAppleStock.recomputations()).toEqual(2);
  });

  it('uses a process function', () => {
    const getStocksSelector = renorm(getStocks, Schemas.COMPANY_ARRAY, {
      process: (result) => decamelizeKeys(result),
    });
    const result = getStocksSelector(state);
    expect(result[1]).toEqual(
      expect.objectContaining({
        company_name: 'Apple Inc.',
        financial_status: 'N',
        id: 'COMP_AAPL',
        market_category: 'Q',
        round_lot_size: 100,
        security_name: 'Apple Inc. - Common Stock',
      })
    );
    expect(result[1]).toMatchSnapshot();

    const getAppleStock = renorm(() => 'AAPL', Schemas.STOCK, {
      process: (result) => decamelizeKeys(result),
    });
    const appleStock = getAppleStock(state);
    expect(appleStock).toMatchSnapshot();
    expect(appleStock).toEqual(
      expect.objectContaining({
        change: -7.08,
        field4: '',
        id: 'AAPL',
        last_earnings_report: {
          earnings: 77598,
          quarter: 3,
          report_id: 'AAPL_QUARTER_3',
        },
      })
    );
  });
});
