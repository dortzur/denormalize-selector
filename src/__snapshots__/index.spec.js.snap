// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create-denormalize-select denormalizes a single entity 1`] = `
Array [
  Object {
    "change": -7.08,
    "field4": "",
    "id": "AAPL",
    "lastEarningsReport": Object {
      "earnings": 77598,
      "quarter": 3,
      "reportId": "AAPL_QUARTER_3",
    },
    "name": "Apple Inc",
    "percentChange": "-4.10%",
    "price": 734,
    "quarterEarnings": Array [
      Object {
        "earnings": 91507,
        "quarter": 1,
        "reportId": "AAPL_QUARTER_1",
      },
      Object {
        "earnings": 69667,
        "quarter": 2,
        "reportId": "AAPL_QUARTER_2",
      },
      Object {
        "earnings": 77598,
        "quarter": 3,
        "reportId": "AAPL_QUARTER_3",
      },
    ],
  },
]
`;

exports[`create-denormalize-select denormalizes entities 1`] = `
Object {
  "companyName": "American Airlines Group, Inc.",
  "financialStatus": "N",
  "id": "COMP_AAL",
  "marketCategory": "Q",
  "roundLotSize": 100,
  "securityName": "American Airlines Group, Inc. - Common Stock",
  "stock": Object {
    "change": -0.1,
    "field4": "",
    "id": "AAL",
    "lastEarningsReport": Object {
      "earnings": 64572,
      "quarter": 4,
      "reportId": "AAL_QUARTER_4",
    },
    "name": "American Airlines Group Inc",
    "percentChange": "-0.21%",
    "price": 46.78,
    "quarterEarnings": Array [
      Object {
        "earnings": 74239,
        "quarter": 1,
        "reportId": "AAL_QUARTER_1",
      },
      Object {
        "earnings": 88368,
        "quarter": 2,
        "reportId": "AAL_QUARTER_2",
      },
      Object {
        "earnings": 79974,
        "quarter": 3,
        "reportId": "AAL_QUARTER_3",
      },
      Object {
        "earnings": 64572,
        "quarter": 4,
        "reportId": "AAL_QUARTER_4",
      },
    ],
  },
  "testIssue": "N",
}
`;

exports[`create-denormalize-select denormalizes entities 2`] = `
Object {
  "companyName": "Adobe Systems Incorporated",
  "financialStatus": "N",
  "id": "COMP_ADBE",
  "marketCategory": "Q",
  "roundLotSize": 100,
  "securityName": "Adobe Systems Incorporated - Common Stock",
  "stock": Object {
    "change": -2.38,
    "field4": "",
    "id": "ADBE",
    "lastEarningsReport": Object {
      "earnings": 61579,
      "quarter": 3,
      "reportId": "ADBE_QUARTER_3",
    },
    "name": "Adobe Systems Inc",
    "percentChange": "-1.05%",
    "price": 225.24,
    "quarterEarnings": Array [
      Object {
        "earnings": 64269,
        "quarter": 1,
        "reportId": "ADBE_QUARTER_1",
      },
      Object {
        "earnings": 65221,
        "quarter": 2,
        "reportId": "ADBE_QUARTER_2",
      },
      Object {
        "earnings": 61579,
        "quarter": 3,
        "reportId": "ADBE_QUARTER_3",
      },
    ],
  },
  "testIssue": "N",
}
`;

exports[`create-denormalize-select denormalizes entities 3`] = `
Object {
  "companyName": "Biogen Idec Inc.",
  "financialStatus": "N",
  "id": "COMP_BIIB",
  "marketCategory": "Q",
  "roundLotSize": 100,
  "securityName": "Biogen Idec Inc. - Common Stock",
  "stock": Object {
    "change": -3,
    "field4": "",
    "id": "BIIB",
    "lastEarningsReport": Object {
      "earnings": 74375,
      "quarter": 4,
      "reportId": "BIIB_QUARTER_4",
    },
    "name": "Biogen Inc",
    "percentChange": "-1.13%",
    "price": 263.02,
    "quarterEarnings": Array [
      Object {
        "earnings": 86549,
        "quarter": 1,
        "reportId": "BIIB_QUARTER_1",
      },
      Object {
        "earnings": 93205,
        "quarter": 2,
        "reportId": "BIIB_QUARTER_2",
      },
      Object {
        "earnings": 95189,
        "quarter": 3,
        "reportId": "BIIB_QUARTER_3",
      },
      Object {
        "earnings": 74375,
        "quarter": 4,
        "reportId": "BIIB_QUARTER_4",
      },
    ],
  },
  "testIssue": "N",
}
`;

exports[`create-denormalize-select denormalizes entities 4`] = `
Object {
  "companyName": "Discovery Communications, Inc.",
  "financialStatus": "N",
  "id": "COMP_DISCK",
  "marketCategory": "Q",
  "roundLotSize": 100,
  "securityName": "Discovery Communications, Inc. - Series C Common Stock",
  "stock": Object {
    "change": -0.17,
    "field4": "",
    "id": "DISCK",
    "lastEarningsReport": Object {
      "earnings": 96306,
      "quarter": 4,
      "reportId": "DISCK_QUARTER_4",
    },
    "name": "Discovery Inc",
    "percentChange": "-0.79%",
    "price": 21.46,
    "quarterEarnings": Array [
      Object {
        "earnings": 78343,
        "quarter": 1,
        "reportId": "DISCK_QUARTER_1",
      },
      Object {
        "earnings": 72267,
        "quarter": 2,
        "reportId": "DISCK_QUARTER_2",
      },
      Object {
        "earnings": 80546,
        "quarter": 3,
        "reportId": "DISCK_QUARTER_3",
      },
      Object {
        "earnings": 96306,
        "quarter": 4,
        "reportId": "DISCK_QUARTER_4",
      },
    ],
  },
  "testIssue": "N",
}
`;
